### ISSUE TITLE
Model performance

### ISSUE THREAD

**Issue by FlyToYourMooN** [mindslab-ai/nuwave2#6]
Hello, could you please tell me how to reproduce the SNR and LSD indicators in the paper? When I used the model to infer all the audio of the eight speakers in the test set, I found that there were some gaps between the SNR and LSD indicators and the paper. Maybe you randomly selected the clips instead of testing all of them?

**Comment by Biqigubafan** [comment]
`from lightning_model import NuWave2
from omegaconf import OmegaConf as OC
import os
import argparse
import datetime
from glob import glob
import torch
import librosa as rosa
from scipy.io.wavfile import write as swrite
import matplotlib.pyplot as plt
from utils.stft import STFTMag
import numpy as np
from scipy.signal import sosfiltfilt
from scipy.signal import butter, cheby1, cheby2, ellip, bessel
from scipy.signal import resample_poly
import random
import soundfile as sf 


def save_stft_mag(wav, fname):
    fig = plt.figure(figsize=(9, 3))
    plt.imshow(rosa.amplitude_to_db(stft(wav[0].detach().cpu()).numpy(),
               ref=np.max, top_db = 80.),
               aspect='auto',
               origin='lower',
               interpolation='none')
    plt.colorbar()
    plt.xlabel('Frames')
    plt.ylabel('Channels')
    plt.tight_layout()
    fig.savefig(fname, format='png')
    plt.close()
    return

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('-c',
                        '--checkpoint',
                        type=str,
                        required=True,
                        help="Checkpoint path")
    parser.add_argument('-i',
                        '--wav',
                        type=str,
                        default=None,
                        help="audio")
    parser.add_argument('--sr',
                        type=int,
                        required=True,
                        help="Sampling rate of input audio")
    parser.add_argument('--steps',
                        type=int,
                        required=False,
                        help="Steps for sampling")
    parser.add_argument('--gt', action="store_true",
                        required=False, help="Whether the input audio is 48 kHz ground truth audio.")
    parser.add_argument('--device',
                        type=str,
                        default='cuda',
                        required=False,
                        help="Device, 'cuda' or 'cpu'")

    args = parser.parse_args()
    #torch.backends.cudnn.benchmark = False
    hparams = OC.load('hparameter.yaml')
    os.makedirs(hparams.log.test_result_dir, exist_ok=True)
    if args.steps is None or args.steps == 8:
        args.steps = 8
        noise_schedule = eval(hparams.dpm.infer_schedule)
    else:
        noise_schedule = None
    model = NuWave2(hparams).to(args.device)
    model.eval()
    stft = STFTMag()
    ckpt = torch.load(args.checkpoint, map_location='cpu')
    model.load_state_dict(ckpt['state_dict'] if not('EMA' in args.checkpoint) else ckpt)

    highcut = args.sr // 2
    nyq = 0.5 * hparams.audio.sampling_rate
    hi = highcut / nyq
    path_ori="/data2/backup/workspace/liukaiyang/tts/vctk_test_flac"
    out_dir="/data/workspace/liukaiyang/TTS/nuwave2-vctk/nuwave2-master/test_wav_24K"
    wav_list=os.listdir(path_ori)
    #wav_list=[i for i in wav_list if i[-1]=="v"]
    os.makedirs("/data/workspace/liukaiyang/TTS/nuwave2-vctk/nuwave2-master/test_wav_24K/ori",exist_ok=True)
    for fwav in wav_list:
        print(fwav)
        wav, _ = rosa.load(os.path.join(path_ori,fwav), sr=hparams.audio.sampling_rate, mono=True)
        wav /= np.max(np.abs(wav))
        wav = wav[:len(wav) - len(wav) % hparams.audio.hop_length]
        sf.write(os.path.join(out_dir,"ori",fwav),wav,hparams.audio.sampling_rate)
        order = 8
        sos = cheby1(order, 0.05, hi, btype='lowpass', output='sos')
        wav_l = sosfiltfilt(sos, wav)

        # downsample to the low sampling rate
        wav_l = resample_poly(wav_l, highcut * 2, hparams.audio.sampling_rate)
        # upsample to the original sampling rate
        wav_l = resample_poly(wav_l, hparams.audio.sampling_rate, highcut * 2)

        if len(wav_l) < len(wav):
            wav_l = np.pad(wav, (0, len(wav) - len(wav_l)), 'constant', constant_values=0)
        elif len(wav_l) > len(wav):
            wav_l = wav_l[:len(wav)]


        fft_size = hparams.audio.filter_length // 2 + 1
        band = torch.zeros(fft_size, dtype=torch.int64)
        band[:int(hi * fft_size)] = 1

        wav = torch.from_numpy(wav).unsqueeze(0).to(args.device)
        wav_l = torch.from_numpy(wav_l.copy()).float().unsqueeze(0).to(args.device)
        band = band.unsqueeze(0).to(args.device)

        wav_recon, wav_list = model.inference(wav_l, band, args.steps, noise_schedule)

        wav = torch.clamp(wav, min=-1, max=1 - torch.finfo(torch.float16).eps)

        wav_l = torch.clamp(wav_l, min=-1, max=1 - torch.finfo(torch.float16).eps)

        wav_recon = torch.clamp(wav_recon, min=-1, max=1 - torch.finfo(torch.float16).eps)
        swrite(os.path.join(hparams.log.test_result_dir, f'result.wav'),
            hparams.audio.sampling_rate, wav_recon[0].detach().cpu().numpy())
        `sf.write(os.path.join(out_dir,fwav),wav_recon[0].detach().cpu().numpy(),hparams.audio.sampling_rate)``

**Comment by FlyToYourMooN** [comment]
![image](https://user-images.githubusercontent.com/32425101/190944829-3cf708b9-dbf2-4765-a7fb-a465337293df.png)

**Comment by FlyToYourMooN** [comment]
As the code says, I still follow your inference steps, just save the original audio and the upsampled audio, and then use your code to calculate the SNR and LSD. The difference between the LSD and the paper is very small, but the SNR will be worse, for example, 24K to 48K, 28.4 in the paper, but the result I get is 24.84

**Comment by Seungwoo0326** [comment]
@Biqigubafan 
Hi, sorry for the late response.
I evaluated our model using the `for_test.py` file. The difference between this code and your process is clipping. In the `for_test.py` file, we evaluate the samples without clipping between -1 to 1. And in your process, the samples are clipped in inference code and then evaluate.
Your way might be correct but I think it doesn't matter because we evaluated all the baseline models in the same way. Anyway, thank you for reporting!
